File: backend/src/db/models/Leaderboard.ts
import { Schema, model, Document } from 'mongoose';

export interface IProblemBreakdown {
  score: number;
  attempted: boolean;
  solvedTime?: string;
}

export interface ILeaderboard extends Document {
  contestId: string;
  userId: string;
  username: string;
  solvedCount: number;
  totalScore: number;
  penaltyTimeMinutes: number;
  problemBreakdown: Map<string, IProblemBreakdown>;
  lastUpdated: Date;
}

const leaderboardSchema = new Schema<ILeaderboard>(
  {
    contestId: { type: String, required: true, ref: 'Contest' },
    userId: { type: String, required: true, ref: 'User' },
    username: { type: String, required: true },
    solvedCount: { type: Number, default: 0 },
    totalScore: { type: Number, default: 0 },
    penaltyTimeMinutes: { type: Number, default: 0 },
    problemBreakdown: {
      type: Map,
      of: new Schema<IProblemBreakdown>(
        {
          score: { type: Number, default: 0 },
          attempted: { type: Boolean, default: false },
          solvedTime: { type: String },
        },
        { _id: false },
      ),
      default: {},
    },
    lastUpdated: { type: Date, default: Date.now },
  },
  {
    timestamps: false,
    versionKey: false,
  },
);

// Compound unique index on (contestId, userId)
leaderboardSchema.index({ contestId: 1, userId: 1 }, { unique: true });

export const Leaderboard = model<ILeaderboard>('Leaderboard', leaderboardSchema);

File: backend/app/models/leaderboard.py
from beanie import Document
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Dict, Optional, Any

class ProblemScoreBreakdown(BaseModel):
    score: int = 0
    attempted: bool = False
    solved_time: Optional[str] = None

class Leaderboard(Document):
    contest_id: str
    user_id: str
    username: str
    solved_count: int = 0
    total_score: int = 0
    penalty_time_minutes: int = 0
    problem_breakdown: Dict[str, ProblemScoreBreakdown] = Field(default_factory=dict)
    last_updated: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "leaderboards"
        indexes = [
            [("contest_id", 1), ("user_id", 1)],
            [("contest_id", 1), ("total_score", -1), ("penalty_time_minutes", 1)],
        ]
